rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,149 @@
+- Start Date: 2020-03-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Applications should be aware when their UI is rendered for purposes of
+capturing a screenshot. This ability would improve the quality of the Kibana
+Reporting feature for a few reasons:
+ - Fewer objects in the headless browser memory since interactive code doesn't run
+ - Fewer Reporting bugs in releases since App teams have more ownership and
+   control over the reportability of their UI
+
+**Screenshot mode service**
+
+Applications in Kibana should be aware of when their rendering is for a human
+user using the page interactively, and when it's not. Sometimes the page is
+loaded for display-purposes: there is no user in control of the
+browser and no need for interactivity. Having Kibana support that purpose
+allows higher-level features to be built on top of rapidly-available images of
+Kibana pages. This turns out to be crucial for features like scheduled PDF
+report generation, and automated alerts with embedded images.
+
+As a bridge between the Kibana applications screenshot-capture features is
+needed as a service in Kibana. This RFC names that service the Screenshot
+Mode Service (name TBD). Applications would read from this service while
+loading UI components into the browser, and help the application make
+case-by-case bases for loading components and the parameters that get passed
+into the components.
+
+In most cases, the information coming from this service would help the
+application determine which UI components should **not** be loaded, as applications will
+use the service response for skipping components that aren't needed for
+presenting the data in a screenshot, and thus work towards better performance
+for the display-only purpose.
+
+More background on how Reporting currently works, including the lifecycle of
+creating a PNG report, is here: https://github.com/elastic/kibana/issues/59396
+
+# Basic example
+
+When Kibana loads initially, there is a Newsfeed component in the UI that
+checks internally cached records to see if it must fetch the Elastic News
+Service for newer items. When the Screenshot Mode Service is implemented, the
+Newsfeed component has a source of information to check on whether or not it
+should load in the Kibana UI. If it can avoid loading, it avoids an unnecessary
+HTTP round trip, which weigh heavily on performance.
```

Comment:
I'm not seeing how a flag in `kibana.json` would solve it. The Service could have a server-side API, but I'm not sure how much plugin logic is involved with server-side rendering.

Reply:
AFAIK, the rendering service loads all the bundles and calls bootstrap:
https://github.com/elastic/kibana/blob/d0949121c82d70c07acc529772eff2738f085b8d/src/core/server/rendering/bootstrap/render_template.ts#L123-L127

The bundles are identified by those plugins with `kibana.json` containing `ui: true`. I think that we could explore the possibility of filtering out the bundles that are not required to load in the screenshot/headless modes or, at least, not calling the `start` method lifecycle for those ones. 

Maybe someone else in the @elastic/kibana-core can verify that I'm not going too crazy here 😅

Reply:
IMO this seems quite risky. Plugins can do all sorts of things that affect the UI and if we did this, we'd need to test rendering in this mode and compare against the normal user mode in order to be sure that nothing significant changes in the rendered output. This sounds pretty hard & time consuming to me, and we still have testing gaps in just normal application usage patterns.

It could be an optimization we explore later, but I suspect the costs of this sort of approach to outweigh the benefits right now. Once we have a more sophisticated testing suite and performance testing, this will be much easier to tackle.